### PR TITLE
Feat: Add Data Quality Modules

### PR DESCRIPTION
**Pull request by lorraine-houston-ensono**
This PR adds data quality functionality to the existing pysparkle library. 

In this first version a stationary config is passed in the entrypoint file in the following version once the deployment mechanism is available this will be abstracted and uploaded as a yaml file

### CODE REVIEW

File: pysparkle/pysparkle/data_quality/data_quality.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+from pyspark.sql import SparkSession
+
+from pysparkle.adls_utils import set_env, set_spark_properties
+from pysparkle.data_quality.data_quality_utils import create_datasource_config, create_expectation_suite, \
+execute_validations
+
+def data_quality_main(DQ_conf):
+    dataset_name = DQ_conf["dataset_name"]
+    print("Running Data Quality processing...")
```
Comment (motylp) [inline pysparkle/pysparkle/data_quality/data_quality.py:9]: please update prints to logger when logging PR is merged

File: pysparkle/pysparkle/data_quality/data_quality.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    spark = SparkSession.builder.appName(
+        f'Data Quality {DQ_conf["container_name"]} {dataset_name}'
```
Comment (motylp) [inline pysparkle/pysparkle/data_quality/data_quality.py:11]: Can you make the application name without whitespaces?
```suggestion
        f'DataQuality-{DQ_conf["container_name"]}-{dataset_name}'
```

File: pysparkle/pysparkle/data_quality/data_quality.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+from pyspark.sql import SparkSession
+
+from pysparkle.adls_utils import set_env, set_spark_properties
+from pysparkle.data_quality.data_quality_utils import create_datasource_config, create_expectation_suite, \
+execute_validations
+
+def data_quality_main(DQ_conf):
```
Comment (motylp) [inline pysparkle/pysparkle/data_quality/data_quality.py:7]: Can you update the argument name to `dq_conf`? (Python naming convention)

File: pysparkle/pysparkle/data_quality/data_quality.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    print("Running Data Quality processing...")
+    spark = SparkSession.builder.appName(
+        f'Data Quality {DQ_conf["container_name"]} {dataset_name}'
+    ).getOrCreate()
+
+    set_env()
+    set_spark_properties(spark)
+
+    table_name = f'{DQ_conf["container_name"]}.{dataset_name}'
+    dm_data = spark.read.table(table_name)
```
Comment (motylp) [inline pysparkle/pysparkle/data_quality/data_quality.py:18]:
```suggestion
    df = spark.read.table(table_name)
```
I don't know what `dm` stands for.

File: pysparkle/pysparkle/data_quality/data_quality.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    gx_dataframe = create_datasource_config(
```
Comment (motylp) [inline pysparkle/pysparkle/data_quality/data_quality.py:19]: the name of this variable is quite misleading - it says it's a dataframe, while it's a context. Maybe rename it to gx_context or gx_config?

File: pysparkle/pysparkle/data_quality/data_quality.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        DQ_conf["dataset_name"], DQ_conf["expectation_suite_name"]
+    )
+
+    gx_dataframe = create_expectation_suite(gx_dataframe, DQ_conf)
```
Comment (motylp) [inline pysparkle/pysparkle/data_quality/data_quality.py:23]: rename to gx_context?

File: pysparkle/pysparkle/data_quality/data_quality_utils.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+def create_datasource_config(datasource_name: str) -> BaseDataContext:
+    """
+    Given a string containing the datasource name, this function generates a data context instance
+
+    Args:
+        datasource_name (str): Name of the datasource to be validated
+
+    Returns:
+        BaseDataContext: populated data context instance to which expectations can be added
```
Comment (motylp) [inline pysparkle/pysparkle/data_quality/data_quality_utils.py:25]:
```suggestion
        Populated data context instance to which expectations can be added
```
Type is already given as a hint in the function definition.

File: pysparkle/pysparkle/data_quality/data_quality_utils.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+) -> ExpectationSuite:
+    """Add expectations for columns as defined in the config file.
+    
+        Args:
+        expectation_suite (ExpectationSuite): Existing expectation suite to be added to
+        validation_conf (dict): dict containing details of validators to be added to columns
+
+    Returns:
+        ExpectationSuite: Expectation suite with new expectations saved
+    """
```
Comment (motylp) [inline pysparkle/pysparkle/data_quality/data_quality_utils.py:74]:
```suggestion
    """Add expectations for columns as defined in the config file.
    
    Args:
        expectation_suite: Existing expectation suite to be added to
        validation_conf: dict containing details of validators to be added to columns

    Returns:
        Expectation suite with new expectations saved
    """
```
Same as in the other place: I'd rather stick to one convention: if we are already using typehints in the function definition, there is no point in repeating them in the docstrings. Docs generation/rendering tools can use the typehints.

File: pysparkle/pysparkle/data_quality/data_quality_utils.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+    return context
+
+def execute_validations(
+        context: BaseDataContext, 
+        DQ_conf: Dict, 
+        df: DataFrame,
+    ) -> ExpectationSuiteValidationResult:
+    """
+    Given a great expectations data context, the relevant config, and a dataframe containing the
```
Comment (motylp) [inline pysparkle/pysparkle/data_quality/data_quality_utils.py:117]:
```suggestion
    Given a Great Expectations data context, the relevant config, and a dataframe containing the
```

File: pysparkle/pysparkle/pysparkle_cli.py [hunk 10]
```diff
@@ -26,8 +27,28 @@ def gold(partitions):
     gold_main(partitions)
 
 
+dq_config = {
```
Comment (motylp) [inline pysparkle/pysparkle/pysparkle_cli.py:30]: I guess it is only temporary to store DQ config here?

Comment (motylp) [reply pysparkle/pysparkle/pysparkle_cli.py:30]: I've just read the description - resolving.

File: pysparkle/pysparkle/data_quality/data_quality.py (hunk 3 above)
Comment (lorraine-houston-ensono) [reply pysparkle/pysparkle/data_quality/data_quality.py:7]: This is going to change to be conf path here but I'll change it elsewhere no problem

File: pysparkle/pysparkle/data_quality/data_quality.py (hunk 4 above)
Comment (lorraine-houston-ensono) [reply pysparkle/pysparkle/data_quality/data_quality.py:18]: Fair shout

File: pysparkle/pysparkle/data_quality/data_quality_utils.py (hunk 8 above)
Comment (lorraine-houston-ensono) [reply pysparkle/pysparkle/data_quality/data_quality_utils.py:74]: Fine by me

File: pysparkle/pysparkle/data_quality/data_quality.py (hunk 6 above)
Comment (lorraine-houston-ensono) [reply pysparkle/pysparkle/data_quality/data_quality.py:23]: will do

File: pysparkle/pysparkle/data_quality/data_quality_utils.py [hunk 11]
```diff
@@ -0,0 +1,139 @@
+"""
+Data_quality utility functions to set up validations
+"""
+from typing import List, Dict
+from pyspark.sql import DataFrame
+
+from great_expectations.core.batch import RuntimeBatchRequest
```
Comment (motylp) [inline pysparkle/pysparkle/data_quality/data_quality_utils.py:7]: Don't we need to add great_expectations to pysparkle's pyproject.toml?

File: pysparkle/pysparkle/data_quality/data_quality_utils.py [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+def add_expectations_for_columns(
+    expectation_suite: ExpectationSuite,
+    validation_conf: Dict,
+) -> ExpectationSuite:
+    """
+    Add expectations for columns as defined in the config file.
+    
+    Args:
+        expectation_suite (ExpectationSuite): Existing expectation suite to be added to
+        validation_conf (dict): dict containing details of validators to be added to columns
```
Comment (motylp) [inline pysparkle/pysparkle/data_quality/data_quality_utils.py:71]:
```suggestion
        expectation_suite: Existing expectation suite to be added to
        validation_conf: dict containing details of validators to be added to columns
```

File: pysparkle/pysparkle/data_quality/data_quality_utils.py [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+def create_expectation_suite(
+    context: BaseDataContext,
+    dq_conf: Dict,
+) -> BaseDataContext:
+    """
+    Creates an expectation suite, and adds expectations to it
+
+    Args:
+        context (BaseDataContext): Existing expectation suite to be added to
+        dq_conf (dict): dict containing details of validators to be added to column
```
Comment (motylp) [inline pysparkle/pysparkle/data_quality/data_quality_utils.py:99]:
```suggestion
        context: Existing expectation suite to be added to
        dq_conf: dict containing details of validators to be added to column
```
here and in other places as in [Example Google Style Python Docstrings](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) - function_with_pep484_type_annotations

**Review by motylp — APPROVED** [review]

**Comment by motylp** [comment]
LGTM - suggested docstring updates according to Google Style Python

**Review by motylp — DISMISSED** [review]

**Review by motylp — DISMISSED** [review]

File: pyproject.toml [hunk 14]
```diff
@@ -14,6 +14,7 @@ azure-identity = "^1.12.0"
 azure-mgmt-datafactory = "^3.1.0"
 azure-storage-file-datalake = "^12.11.0"
 polling2 = "^0.5.0"
+great-expectations = "^0.15.15"
```
Comment (motylp) [inline pyproject.toml:17]: Why do we need GX in the top level pyproject.toml?

Comment (lorraine-houston-ensono) [reply pyproject.toml:17]: We don't, I got in a battle with the lock files and it's clearly ended up in the wrong place I'll remove it

File: pysparkle/pysparkle/data_quality/data_quality_utils.py [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+        context: BaseDataContext, 
+        dq_conf: Dict, 
+        df: DataFrame,
+    ) -> ExpectationSuiteValidationResult:
+    """
+    Given a Great Expectations data context, the relevant config, and a dataframe containing the
+    data to be validated. This function runs the validations and returns the result
+
+    Args:
+        validator (Validator): Great Expectations validator object for the given datamart
```
Comment (motylp) [inline pysparkle/pysparkle/data_quality/data_quality_utils.py:125]: please adjust args

Comment (lorraine-houston-ensono) [reply pysparkle/pysparkle/data_quality/data_quality_utils.py]: Done

File: pysparkle/pysparkle/data_quality/data_quality_utils.py [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+    return context
+
+def execute_validations(
+        context: BaseDataContext, 
+        dq_conf: Dict, 
+        df: DataFrame,
+    ) -> ExpectationSuiteValidationResult:
+    """
+    Given a Great Expectations data context, the relevant config, and a dataframe containing the
+    data to be validated. This function runs the validations and returns the result
```
[the remark this answers is not in the document]
Comment (lorraine-houston-ensono) [reply pysparkle/pysparkle/data_quality/data_quality_utils.py]: Done

**Review by motylp — APPROVED** [review]

File: pysparkle/pysparkle/data_quality/data_quality.py [hunk 17]
```diff
@@ -0,0 +1,35 @@
+import logging
```
Comment (motylp) [inline pysparkle/pysparkle/data_quality/data_quality.py:1]: I think __init__.py is missing for data_quality package - the app is failing while importing

Comment (lorraine-houston-ensono) [reply pysparkle/pysparkle/data_quality/data_quality.py:1]: Yup writing the unit tests now so caught this and a couple of other things. The pipeline run issue won't be fixed till this afternoon so I'll just put the unit tests and fixes in this PR aswell. I'll ping you when it's ready for another look @motylp

File: pysparkle/pysparkle/data_quality/data_quality_utils.py [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+from great_expectations.data_context.types.base import DataContextConfig, \
+    FilesystemStoreBackendDefaults
+
+
+def create_datasource_config(datasource_name: str) -> BaseDataContext:
+    """
+    Given a string containing the datasource name, this function generates a data context instance
+
+    Args:
+        datasource_name: Name of the datasource to be validated
```
Comment (motylp) [inline pysparkle/pysparkle/data_quality/data_quality_utils.py:22]: Can you explain some more what is data source?

File: pysparkle/pysparkle/data_quality/data_quality_utils.py [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+    )
+    expectation_suite = add_expectations_for_columns(
+        expectation_suite,
+        dq_conf["validation_config"]
+    )
+    context.save_expectation_suite(expectation_suite)
+
+    return context
+
+def execute_validations(
```
Comment (motylp) [inline pysparkle/pysparkle/data_quality/data_quality_utils.py:115]: PyCharm is complaining (here and in some other similar places): PEP 8: E302 expected 2 blank lines, found 1

Can you pls try to make formatting consistent? https://peps.python.org/pep-0008/#blank-lines

File: pysparkle/pysparkle/data_quality/data_quality_utils.py (hunk 18 above)
Comment (lorraine-houston-ensono) [reply pysparkle/pysparkle/data_quality/data_quality_utils.py:22]: https://docs.greatexpectations.io/docs/terms/datasource/ So generally we'll likely use the raw data source or table name here, we're inferring the schema anyway but it needs to be unique for the validators. This is standard GX language as above but we basically use it to generate the other required names.

File: pysparkle/pysparkle/pysparkle_cli.py [hunk 20]
```diff
@@ -37,8 +38,34 @@ def gold(partitions):
...
+    "container_name": "staging",
+    "dataset_name": "movies_metadata",
```
Comment (motylp) [inline pysparkle/pysparkle/pysparkle_cli.py:43]: So currently in bronze (`raw`) container we have `movies_dataset` "folder" and inside this folder there is a number of CSV files. All these CSV files are written directly to "staging" schema. Aren't we mixing the dataset name with table names?

Comment (motylp) [reply pysparkle/pysparkle/pysparkle_cli.py:43]: Yeah, I was initially confused by naming it "dataset_name" - while you were replying I edited my initial comment to simply change this parameter name if possible:

```suggestion
    "table_name": "movies_metadata",
```

Comment (lorraine-houston-ensono) [reply pysparkle/pysparkle/pysparkle_cli.py:43]: Can do, I'd just gone with dataset as it's the GX standard language, but happy to amend this to table_name if it's more user friendly. I'd bear in mind though that we'll use the same config model regardless of raw/staging etc so I'm not sure table name will always make sense (although for the same reason dataset_name might not either

Comment (motylp) [reply pysparkle/pysparkle/pysparkle_cli.py:43]: Ok for me to rename the other one to `folder_name` in a separate PR?

Comment (motylp) [reply pysparkle/pysparkle/pysparkle_cli.py:43]: yes, I am! :D

File: pysparkle/pysparkle/data_quality/data_quality_utils.py (hunk 18 above)
Comment (motylp) [reply pysparkle/pysparkle/data_quality/data_quality_utils.py:25]: Thanks, I think adding this URL to the arg description would be helpful, but if you think this is obvious, feel free to resolve.

File: pysparkle/pysparkle/pysparkle_cli.py (hunk 20 above)
Comment (lorraine-houston-ensono) [reply pysparkle/pysparkle/pysparkle_cli.py:43]: Done

File: pysparkle/pysparkle/data_quality/data_quality_utils.py (hunk 18 above)
Comment (lorraine-houston-ensono) [reply pysparkle/pysparkle/data_quality/data_quality_utils.py:25]: We're using datasource name everywhere now so I'd be putting the URL in alot of places, I think leave it for now and then we can add it if needed in future

File: pysparkle/pysparkle/data_quality/data_quality_utils.py [hunk 21]
```diff
@@ -0,0 +1,153 @@
...
+                    },
+                )
+            )
+
+    return expectation_suite
+
+
+def create_expectation_suite(
+    context: BaseDataContext,
+    dq_conf: Dict,
```
[the remark this answers is not in the document]
Comment (lorraine-houston-ensono) [reply pysparkle/pysparkle/data_quality/data_quality_utils.py:97]: Can this be added at a later date? This PR is growing arms and legs and already has more work that will be added shortly.

[the remark this answers is not in the document]
Comment (motylp) [reply pysparkle/pysparkle/data_quality/data_quality_utils.py:97]: Yeah, sure, how about a @TODO comment to refactor it in the future? Feel free to comment or not and resolve :)

File: pysparkle/pysparkle/pysparkle_cli.py (hunk 20 above)
Comment (motylp) [inline pysparkle/pysparkle/pysparkle_cli.py:43]:
```suggestion
    "datasource_name": "movies_metadata",
```

Comment (lorraine-houston-ensono) [reply pysparkle/pysparkle/pysparkle_cli.py]: done

File: pysparkle/tests/unit/data_quality/test_data_quality_utils.py [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+                    "expectation_kwargs": {},
+                }
+            ],
+        }
+    ],
+}
+
+
+def test_create_datasource_context():
+    context = create_datasource_context(TEST_DQ_CONF["datasource_name"], TEST_DQ_CONF["gx_directory_path"])
```
Comment (motylp) [inline pysparkle/tests/unit/data_quality/test_data_quality_utils.py:31]: I would rather have a fixture with creating a datasource_context that would create it using pytest's `tmp_path` fixture. That way we wouldn't have to bother with removing this directory. Right now, if the test fails, it won't be removed. But I'm OK to do it in the next round. If you agree, could you add a task to one of the DQ stories?

Comment (lorraine-houston-ensono) [reply pysparkle/tests/unit/data_quality/test_data_quality_utils.py:31]: I've got another PR going in soon anyway I'll do it in that

**Pull request merged**